betatest/tests/browser/browser_registration.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Beta registration browser tests
+"""
+
+# Imports #####################################################################
+
+from django.core.urlresolvers import reverse
+from django.contrib.staticfiles.testing import StaticLiveServerTestCase
+from selenium import webdriver
+from selenium.webdriver.support import expected_conditions
+from selenium.webdriver.support.ui import WebDriverWait
+
+from betatest.tests.test_views import BetaTestApplicationViewMixin
+
+
+# Tests #######################################################################
+
+class BetaTestBrowserTestCase(BetaTestApplicationViewMixin,
+                              StaticLiveServerTestCase):
+    """
+    Tests the beta test registration flow with a real browser.
+    """
+    def setUp(self):
+        super().setUp()
+        self.client = webdriver.Firefox()
+
+    def tearDown(self):
+        self.client.quit()
+        super().tearDown()
+
+    def register(self, form_data):
+        """
+        Fill in the registration form and click the submit button, if enabled.
+        """
+        self.client.get('{host}{path}'.format(
+            host=self.live_server_url,
+            path=reverse('beta:register'),
+        ))
+        form = self.client.find_element_by_tag_name('form')
+
+        for field, value in form_data.items():
+            element = form.find_element_by_name(field)
+            if (element.get_attribute('type') == 'checkbox' and
+                    bool(value) != element.is_selected()):
+                element.click()
+            else:
+                element.send_keys(value)
+
+        submit = form.find_element_by_tag_name('button')
+        if submit.is_enabled():
+            # Click the submit button and wait for the next page to load
+            html = self.client.find_element_by_tag_name('html')
+            submit.click()
+            WebDriverWait(self.client, timeout=3) \
+                .until(expected_conditions.staleness_of(html))
+
+        return self.client.page_source
+
+    def assert_registration_fails(self, form_data):
```

Comment:
The test methods are defined in the mixin here: https://github.com/open-craft/opencraft/blob/omar/beta-registration/betatest/tests/test_views.py#L39

The code in this class fills in and submits the form using selenium instead of sending the form data directly using the django test client.